File: backend/app/routers/credits.py
"""
SafeVision API - Credits Router
Endpoints for checking credits and creating checkout sessions.
"""

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from app.models import CreditsResponse
from app.services.autumn import autumn_service
from app.middleware.credits import get_customer_id

router = APIRouter(tags=["Credits"])


@router.get(
    "/credits",
    response_model=CreditsResponse,
    summary="Check remaining credits",
    description="Returns your current credit balance and plan information.",
)
async def get_credits(
    customer_id: Optional[str] = Depends(get_customer_id),
):
    if not autumn_service.enabled:
        return CreditsResponse(
            enabled=False,
            customer_id=None,
            credits_remaining=None,
            plan=None,
        )

    if not customer_id:
        raise HTTPException(status_code=401, detail="API key required")

    # Get balance
    credit_info = await autumn_service.check_credits(customer_id)
    customer_info = await autumn_service.get_customer_info(customer_id)

    plan_name = None
    if customer_info and "products" in customer_info:
        products = customer_info["products"]
        if products:
            plan_name = products[0].get("name") if isinstance(products[0], dict) else str(products[0])

    return CreditsResponse(
        enabled=True,
        customer_id=customer_id,
        credits_remaining=credit_info.get("balance"),
        plan=plan_name,
    )


@router.get(
    "/credits/checkout",
    summary="Get checkout URL for purchasing credits",
    description="Returns a Stripe checkout URL to purchase a credit plan.",
)
async def get_checkout_url(
    product_id: str = Query(..., description="Product/plan ID to purchase"),
    customer_id: Optional[str] = Depends(get_customer_id),
):
    if not autumn_service.enabled:
        raise HTTPException(status_code=404, detail="Credit system not enabled")

    if not customer_id:
        raise HTTPException(status_code=401, detail="API key required")

    url = await autumn_service.create_checkout(customer_id, product_id)
    if not url:
        raise HTTPException(status_code=500, detail="Failed to create checkout session")

    return {"checkout_url": url}

File: middleware/src/routes/credits.ts
/**
 * /api/v1/credits — check credit balance
 * /api/v1/credits/checkout — create a checkout session
 */

import { Router, Request, Response } from "express";
import { apiKeyAuth } from "../middleware/auth";
import { autumnService } from "../services/autumn";

export const creditsRouter = Router();

creditsRouter.get("/credits", apiKeyAuth, async (req: Request, res: Response) => {
  const customerId = req.customerId || "anonymous";

  if (!autumnService.isEnabled) {
    return res.json({
      enabled: false,
      customer_id: null,
      credits_remaining: null,
      plan: null,
    });
  }

  const info = await autumnService.getCustomerInfo(customerId);
  const { balance } = await autumnService.checkCredits(customerId);

  return res.json({
    enabled: true,
    customer_id: customerId,
    credits_remaining: balance,
    plan: info?.plan || null,
  });
});

creditsRouter.post(
  "/credits/checkout",
  apiKeyAuth,
  async (req: Request, res: Response) => {
    const customerId = req.customerId || "anonymous";
    const { product_id } = req.body;

    if (!product_id) {
      return res.status(400).json({
        status: "error",
        error: "product_id is required",
      });
    }

    if (!autumnService.isEnabled) {
      return res.status(400).json({
        status: "error",
        error: "Credits are not enabled on this server",
      });
    }

    const url = await autumnService.createCheckout(customerId, product_id);
    if (!url) {
      return res.status(500).json({
        status: "error",
        error: "Could not create checkout session",
      });
    }

    return res.json({ url });
  }
);
